flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/ValueCountCombineReader.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.storage.filestore.combine;
+
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.typeutils.RowDataSerializer;
+import org.apache.flink.table.storage.filestore.KeyValue;
+import org.apache.flink.table.storage.filestore.ValueKind;
+import org.apache.flink.table.storage.filestore.utils.RecordReader;
+
+import java.util.Comparator;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * A value count {@link CombineRecordReader} which sums the count in the value. It only accepts
+ * record with {@link ValueKind#ADD} and count cannot be null.
+ */
+public class ValueCountCombineReader extends CombineRecordReader {
+
+    public ValueCountCombineReader(
+            RecordReader<KeyValue> reader,
+            Comparator<RowData> userKeyComparator,
+            RowDataSerializer userKeySerializer,
+            RowDataSerializer valueSerializer) {
+        super(reader, userKeyComparator, userKeySerializer, valueSerializer);
+    }
+
+    @Override
+    public KeyValue combine(KeyValue from, KeyValue to) {
+        check(from);
+        check(to);
+        return to.setValue(GenericRowData.of(count(to) + count(from)));
+    }
+
+    private void check(KeyValue keyValue) {
+        checkArgument(keyValue.valueKind() == ValueKind.ADD, "Value kind should be ADD.");
+        checkArgument(!keyValue.value().isNullAt(0), "Value count should not be null.");
+    }
+
+    private long count(KeyValue keyValue) {
+        return keyValue.value().getLong(0);
```

Comment:
So `keyValue.value()` is not a pure record? Why not make this `count` as a member of `KeyValue` instead of squishing it in a `RowData`?

Reply:
KeyValue is a common structure, RowData can cover all cases.

Reply:
This is hard for another person to reason about what is stored in `value`. Adding a member for count doesn't hurt. For records with primary keys their count can always be 1.

Reply:
If we add a count to KeyValue.
What is the value? How to serialize KeyValue?

Reply:
OK I understand. For records without primary key the whole record is stored as key and the value is a mere count.

Please add comments in `KeyValue` about these two situations for the ease of understanding.

Reply:
I keep `KeyValue` clean.  "the value is a mere count" is in the user (`ValueCountCombineReader`)